### ISSUE TITLE
The job script generators change too late to the working directory

### ISSUE THREAD

**Issue by vkarak** [eth-cscs/reframe#35]
Quoting from @bcfriesen:

> Actually, now there is a new issue related to this.

> When using DataWarp, the scheduler creates a new temporary directory on the DataWarp file system which is pointed to by an environment variable. For "striped" access, the environment variable is $DW_JOB_STRIPED. The data specified by the #DW stage_in header in the batch script is automatically copied to $DW_JOB_STRIPED. The remainder of the batch script typically involves copying the executable into $DW_JOB_STRIPED, because the value of that environment variable changes with every run (so it can't be hard-coded into the batch script, or into the code itself).

> I tried to address this by writing a simple bash script which copies the executable into $DW_JOB_STRIPED:

> ```#!/bin/bash
> EXE=a.out
> cp $EXE $DW_JOB_STRIPED
> cd $DW_JOB_STRIPED
> ```
> and putting it into the self.job.pre_run stage as described here. The issue is that the auto-generated Slurm script will execute that bash script, but then follow it immediately with cd /path/to/stagedir, as specified by the stagedir in the _site_configuration. So the test never actually tests the functionality of DataWarp because it reads everything from stagedir, which is not on the DataWarp file system.

Yes, this is indeed a problem of our job script generator. The `cd` to the working directory should be done at the very beginning or we should try using the `--workdir` option of Slurm instead.

**Comment by vkarak** [comment]
Hi @bcfriesen, did the workaround proposed in #29 work for you?

**Comment by bcfriesen** [comment]
Hi @vkarak, the workaround in #29 did not work, but at least helped progress the error a little bit farther.

The `LauncherWrapper` that you provided has the expected behavior, i.e., the last line of the generated Slurm script is now
```
cd $DW_JOB_STRIPED &&  srun DW_scratch_stage_in.ex
```

However, the problem now is that the executable still lives in a descendant directory of `stagedir`, so the script fails because `srun` is unable to find the executable in `$DW_JOB_STRIPED`. In normal jobs, one solves this problem by staging in an entire directory which includes the exe itself as well as its inputs. Then when you `cd $DW_JOB_STRIPED`, everything is there. However it's not clear to me how to do that in ReFrame.

**Comment by vkarak** [comment]
I think you can still use the `pre_run` attribute (from 2.10 it has moved to the `RegressionTest` for convenience) to do the initial copy of the executable to the `$DW_JOB_STRIPED` directory. So you can do sth like that:

```python
def __init__(...):
    ...
    self.pre_run = [ 'cp %s $DW_JOB_STRIPED' % self.executable]

def setup(self, partition, environ, **job_opts):
    super().setup(partition, environ, **job_opts)
    self.job.launcher = LauncherWrapper(self.job.launcher, 'cd $DW_JOB_STRIPED && ')
```

Btw, we are now opening the development of the framework, so you can get bug fixes and features as soon as they are merged into master.

**Comment by bcfriesen** [comment]
OK, the `cp` trick should work, thanks. Please note, however, that this approach partially circumvents the real thing that I'm hoping to test, which is the functionality of the `stage_in` directive itself. If we use `cp` then we are checking that DataWarp correctly creates the `$DW_DIR_STRIPED` directory, but we do not check Slurm's ability to stage data into `$DW_DIR_STRIPED` (since staging via `stage_in` uses a different I/O mechanism than plain `cp`).

**Comment by bcfriesen** [comment]
Oh, never mind, I was wrong - the inputs do get staged in with `stage_in`, but the executable does not. So this solution should be fine. Thanks @vkarak.

**Comment by vkarak** [comment]
I was about to answer the same thing! In any case, we're gonna fix the `cd`-ing to the directory, cos it doesn't make sense to do it just before `srun`.

**Comment by bcfriesen** [comment]
OK getting closer, this workaround almost works. The test now has
```
[...]

        self.executable = "./DW_scratch_stage_in.ex"
        self.extra_resources = {
            'datawarp': {
                'capacity': '100GB',
                'mode': 'striped',
                'dw_src': '/path/to/input_data',
                'dw_dst': '$DW_JOB_STRIPED/input_data',
                'stage_filetype': 'file'
            }
        }

[...]

    def setup(self, partition, environ, **job_opts):
        super().setup(partition, environ, **job_opts)
        self.job.pre_run = [ 'cp %s $DW_JOB_STRIPED' % self.executable]
        self.job.launcher = LauncherWrapper(self.job.launcher, 'cd $DW_JOB_STRIPED && ')
```
But since ReFrame submits the Slurm script from a different location than the executable exists (it creates subdirs based on the `PrgEnv` and so on), the `cp` step fails. If I `cd` directly to the directory where the exe and the Slurm script live and submit the Slurm script manually from there, then everything works. So it seems I need to put an absolute dir somewhere in the test script so that `cp` can find the exe, but I don't know where to put it.

**Comment by vkarak** [comment]
Nice, could you copy paste here the generated job script and your desired one? Generally, you can access the absolute path of the stage directory (this path you described above) from the `setup()` phase onward as `self.stagedir`.

**Comment by vkarak** [comment]
I think, I understood the problem. Practically, you want the absolute path of the executable. You can the following line before setting the launcher:

```python
self.executable = os.path.join(self.stagedir, self.executable)
```

**Comment by bcfriesen** [comment]
Here are the relevant bits from the "correct" Slurm script. It will run successfully if and only if the script is submitted from its PWD:
```
module unload PrgEnv-intel
module load PrgEnv-cray
cp ./DW_scratch_stage_in.ex $DW_JOB_STRIPED
cd /path/to/scratch/stage/datawarp_scratch_knl_stage_in/DW_scratch_stage_in/PrgEnv-cray
cd $DW_JOB_STRIPED &&  srun ./DW_scratch_stage_in.ex
```
The way to make it universally "correct" (i.e., submit-able from anywhere) would be to give the target exe its absolute location in the file system, e.g., the last line should become
```
srun $DW_JOB_STRIPED/DW_scratch_stage_in.ex
```

**Comment by vkarak** [comment]
I think I've reproduced the problem with the following small test:

```python
import os
import reframe.utility.sanity as sn

from reframe.core.launchers import LauncherWrapper
from reframe.core.pipeline import RunOnlyRegressionTest


class MyTest(RunOnlyRegressionTest):
    def __init__(self, **kwargs):
        super().__init__('test_check', os.path.dirname(__file__), **kwargs)
        self.valid_systems = ['*']
        self.valid_prog_environs = ['*']
        self.executable = './hello.sh'
        self.sanity_patterns = sn.assert_found('Hello, World', self.stdout)

    def setup(self, partition, environ, **job_opts):
        super().setup(partition, environ, **job_opts)
        self.job.pre_run = ['cp %s $SCRATCH' % self.executable]
        self.job.launcher = LauncherWrapper(self.job.launcher,
                                            'cd $SCRATCH &&')


def _get_checks(**kwargs):
    return [MyTest(**kwargs)]
```

The generated script is the following which matches yours (I am just using `$SCRATCH` and no parallel launcher):

```bash
#!/bin/bash -l
module unload PrgEnv-cray
module load PrgEnv-gnu
cp ./hello.sh $SCRATCH
cd /path/to/stage/login/test_check/PrgEnv-gnu
cd $SCRATCH && ./hello.sh
```

The problem is that ReFrame submits the job script from the directory it was launched by using absolute paths instead of switching to the stage directory. If it would, this example would work nicely. Nonetheless, you can work around it as follows by replacing the `pre_run` assignment with this one:

```python
        self.job.pre_run = ['cp %s $SCRATCH' %
                            os.path.join(self.stagedir, self.executable)]
```

This will generate the following script, which is universally correct (assuming that `$SCRATCH` is an absolute path):

```bash
#!/bin/bash -l
module unload PrgEnv-cray
module load PrgEnv-gnu
cp /path/to/stage/login/test_check/PrgEnv-gnu/./hello.sh $SCRATCH
cd /path/to/stage/login/test_check/PrgEnv-gnu
cd $SCRATCH && ./hello.sh
```

We should consider submitting the jobs from the stage directory to avoid such surprises as we do with the rest of the pipeline phases. Then there will be no need to emit the `cd` command in the job script.

**Comment by vkarak** [comment]
Hi @bcfriesen,

This issue is now being addressed in #128. As soon as this is merged, you will be able to write your test without any workaround and without overriding the `setup()` method:

```python
import os
import reframe.utility.sanity as sn

from reframe.core.pipeline import RunOnlyRegressionTest


class MyTest(RunOnlyRegressionTest):
    def __init__(self, **kwargs):
        super().__init__('test_check', os.path.dirname(__file__), **kwargs)
        self.valid_systems = ['*']
        self.valid_prog_environs = ['*']
        self.executable = './hello.sh'
        self.sanity_patterns = sn.assert_found('Hello, World', self.stdout)
        self.pre_run = ['cp %s $SCRATCH' % self.executable, 'cd $SCRATCH']


def _get_checks(**kwargs):
    return [MyTest(**kwargs)]
```

The generated script is the following, which is also more "natural":

```bash
#!/bin/bash -l
cp ./hello.sh $SCRATCH
cd $SCRATCH
./hello.sh
```
